File: tests/journal.lib/index_cascade.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the journal
#include <pyre/journal.h>
// support
#include <cassert>


// type aliases
// instantiate a functional index
using index_t = pyre::journal::index_t;
// the trash can
using trash_t = pyre::journal::trash_t;


// exercise the cascade feature of the channel state index
int main() {
    // make an index
    index_t index(true, false);

    // lookup a name
    auto & parent = index.lookup("test.index.parent");
    // make sure its activation state is what's expected
    assert(parent.active() == index.active());
    // its fatal state is what's expected
    assert(parent.fatal() == index.fatal());
    // and the device is null
    assert(parent.device() == nullptr);
    // turn it off
    parent.active(false);
    // and make it fatal
    parent.fatal(true);
    // make sure it happened
    assert(parent.active() == false);
    // set the device to a trash can
    parent.device<trash_t>();

    // lookup a name that's lower in the hierarchy
    auto & child = index.lookup("test.index.parent.blah.blah.child");
    // make sure its actual state is what's expected
    assert(child.active() == parent.active());
    // check its fatal state is what's expected
    assert(child.fatal() == parent.fatal());
    // and that it inherited the device
    assert(child.device() == parent.device());

    // all done
    return 0;
}


// end of file

File: tests/journal.pkg/index_cascade.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


def test():
    """
    Verify that channels lower in the hierarchy inherit the default state of their parent
    """
    # get the channel base class
    from journal.Channel import Channel

    # make a severity
    class Severity(Channel, active=True, fatal=True):
        """
        A sample derivation
        """

    # get the trash can
    from journal.Trash import Trash

    # get the index
    index = Severity.index
    # and the inventory type
    inventory_type = Severity.inventory_type

    # look up a name
    parent = index.lookup(name="test.index.parent")
    # verify it is an instance of the correct class
    assert isinstance(parent, inventory_type)
    # hence the state is on
    assert parent.active is True
    # it is fatal
    assert parent.fatal is True
    # and the device is null
    assert parent.device is None

    # deactivate it
    parent.active = False
    # make it non-fatal
    parent.fatal = False
    # and set the device to a trash can
    parent.device = Trash()

    # lookup a name that is lower in the hierarchy
    child = index.lookup(name="test.index.parent.blah.blah.child")
    # make sure it's an instance of the correct type
    assert isinstance(child, inventory_type)
    # that it's state is the same as the parent
    assert child.active == parent.active
    assert child.fatal == parent.fatal
    # and that it inherited the device correctly
    assert child.device is parent.device

    # all done
    return


# main
if __name__ == "__main__":
    # prohibit the journal bindings
    journal_no_libjournal = True
    # run the test
    test()


# end of file
